### PR TITLE
samples: usb: hid-mouse: multiple buttons

### PR DESCRIPTION
**Pull request by masz-nordic**
Added right button, X and Y axis movement to HID mouse example.
Enabled only if board has enough buttons.
Added LED toggling notification when report is sent.

Signed-off-by: Marcin Szymczyk <Marcin.Szymczyk@nordicsemi.no>

### CODE REVIEW

File: samples/subsys/usb/hid-mouse/src/main.c [hunk 1]
```diff
@@ -92,75 +149,178 @@ static const u8_t hid_report_desc[] = {
...
+	}
+
+	if (status[MOUSE_Y_REPORT_POS] != state) {
+		status[MOUSE_Y_REPORT_POS] = state;
+		k_sem_give(&sem);
+	}
+}
+#endif
+
+int configure_callback(struct device *gpio, u32_t pin, int flags,
```
Comment (finikorg) [inline samples/subsys/usb/hid-mouse/src/main.c]: Is this a callback?

File: samples/subsys/usb/hid-mouse/src/main.c [hunk 2]
```diff
@@ -92,75 +149,178 @@ static const u8_t hid_report_desc[] = {
...
-		status = state;
+	if (status[MOUSE_BTN_REPORT_POS] != state) {
+		status[MOUSE_BTN_REPORT_POS] = state;
+		k_sem_give(&sem);
+	}
+}
+
+#ifdef SW1_GPIO_PIN
+static void right_button(struct device *gpio, struct gpio_callback *cb,
+		  u32_t pins)
```
Comment (finikorg) [inline samples/subsys/usb/hid-mouse/src/main.c]: is this line aligned?

**Comment by finikorg** [comment]
You need to add information to readme also

File: samples/subsys/usb/hid-mouse/src/main.c (hunk 1 above)
Comment (masz-nordic) [reply samples/subsys/usb/hid-mouse/src/main.c]: No, it configures GPIOs and adds callbacks to button pins. I see now this might not be the best name. `callbacks_configure()` maybe?

Comment (finikorg) [reply samples/subsys/usb/hid-mouse/src/main.c]: Yes, this sound better, at first it sounds like a callback

File: samples/subsys/usb/hid-mouse/src/main.c [hunk 3]
```diff
@@ -92,75 +149,178 @@ static const u8_t hid_report_desc[] = {
 		/* END_COLLECTION				C0    */
 };
 
-static int get_report_cb(struct usb_setup_packet *setup, s32_t *len,
```
Comment (qianfan-Zhao) [inline samples/subsys/usb/hid-mouse/src/main.c]: Based on our discussion on Slack, this is a bug. I prefer move this fix to another PR.

Comment (qianfan-Zhao) [inline samples/subsys/usb/hid-mouse/src/main.c]: Based on our discussion on Slack, this is a bug. I prefer move this fix to another PR

File: samples/subsys/usb/hid-mouse/src/main.c [hunk 4]
```diff
@@ -18,22 +22,75 @@
...
 #else
 #error SW0_GPIO_FLAGS needs to be set
 #endif
 
+/* If second button exists, use it as right-click. */
+#ifdef SW1_GPIO_PIN
+#define PIN1	SW1_GPIO_PIN
+#endif
+
+#ifdef SW1_GPIO_NAME
```
Comment (qianfan-Zhao) [inline samples/subsys/usb/hid-mouse/src/main.c]: Actually, the first version I wrote, used this marco to detect if a PIN was defined or not. But in that times Zephyr doesn't support write BUTTON node in dts, somethings like SW0_GPIO_CONTROLLER are defined in board.h

But Zephyr support adding BUTTON node in dts now, so this check is useless. I think somethings like https://github.com/zephyrproject-rtos/zephyr/blob/93fc23bc7e458f9155c59980a86226a00b69aa9a/samples/basic/blinky/src/main.c#L11 is enough.

@galak Could you review this sections?

**Review by qianfan-Zhao — CHANGES_REQUESTED** [review]

**Review by dbkinder — CHANGES_REQUESTED** [review]

**Review by pawelzadrozniak — APPROVED** [review]

File: samples/subsys/usb/hid-mouse/src/main.c (hunk 4 above)
Comment (jfischer-phytec-iot) [reply samples/subsys/usb/hid-mouse/src/main.c]: It seems that we can simplify it and use only  SW*_GPIO_CONTROLLER, SW*_GPIO_FLAGS, SW*_GPIO_PIN now

File: samples/subsys/usb/hid-mouse/src/main.c [hunk 5]
```diff
@@ -105,62 +162,174 @@ static struct hid_ops ops = {
 	.get_report = get_report_cb,
 };
 
-static int def_val;
-static volatile int status;
+static u32_t def_val[4];
```
Comment (jfischer-phytec-iot) [inline samples/subsys/usb/hid-mouse/src/main.c]: It may be correct, but not nice to change the type of the variable in this commit.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: samples/subsys/usb/hid-mouse/src/main.c (hunk 3 above)
Comment (masz-nordic) [reply samples/subsys/usb/hid-mouse/src/main.c]: Removed from this PR, I will make a new one specifically for this bugfix.

File: samples/subsys/usb/hid-mouse/src/main.c (hunk 4 above)
Comment (masz-nordic) [reply samples/subsys/usb/hid-mouse/src/main.c]: Removed SW*_GPIO_NAME ifdefs.

File: samples/subsys/usb/hid-mouse/src/main.c (hunk 5 above)
Comment (masz-nordic) [reply samples/subsys/usb/hid-mouse/src/main.c]: Should I make a separate one?

Comment (jfischer-phytec-iot) [reply samples/subsys/usb/hid-mouse/src/main.c]: I think yes, would be more accurate :-)

Comment (masz-nordic) [reply samples/subsys/usb/hid-mouse/src/main.c]: Done :)

**Review by jfischer-no — APPROVED** [review]

File: samples/subsys/usb/hid-mouse/src/main.c [hunk 6]
```diff
@@ -105,63 +147,174 @@ static struct hid_ops ops = {
...
+{
+	u32_t cur_val;
+	u8_t state = status[MOUSE_X_REPORT_POS];
+
+	gpio_pin_read(gpio, PIN2, &cur_val);
+
+	if (def_val[2] != cur_val) {
+		state += 10;
+	} else {
+
```
Comment (qianfan-Zhao) [inline samples/subsys/usb/hid-mouse/src/main.c]: How about remove this unused else?

Comment (masz-nordic) [reply samples/subsys/usb/hid-mouse/src/main.c]: Removed, thanks.

**Review by qianfan-Zhao — APPROVED** [review]

**Comment by masz-nordic** [comment]
@dbkinder @finikorg please recheck

**Review by dbkinder — APPROVED** [review]
LGTM, thanks.

**Review by carlescufi — APPROVED** [review]
